20240328-queries-with-usersets.md

```diff
@@ -0,0 +1,176 @@
+
+Let's assume that we have an FGA model such as the following:
+```
+model
+  schema 1.1
+type employee
+type group
+  relations
+    define member: [employee, group#member]
+```
+
+and the following relationship tuples:
+- `group:eng#member@group:fga#member`
+- `group:fga#member@employee:jon`
+
+The M2G index would be defined as:
+
+| object_type | object_id | relation | subject_type | subject_id | subject_relation |
+|-------------|-----------|----------|--------------|------------|------------------|
+| group       | fga       | member   | employee     | jon        |                  |
+
+And the G2G index, naively, would be defined as:
+
+| object_type | object_id | relation | subject_type | subject_id | subject_relation |
+|-------------|-----------|----------|--------------|------------|------------------|
+| group       | eng       | member   | group        | fga        | member           |
+
+We take the intersection of M2G and G2G, namely
+
+```sql
+SELECT g2g.object_type, g2g.object_id, g2g.relation, m2g.subject_type, m2g.subject_id
+FROM M2G m2g 
+INNER JOIN G2G on m2g.object_type=g2g.object_type AND m2g.relation=g2g.relation
+```
+which returns
+
+| object_type | object_id | relation | subject_type | subject_id |
+|-------------|-----------|----------|--------------|------------|
+| group       | eng       | member   | employee     | jon        |
+
+Notice that we account for the `group:eng#member` relationship for `user:jon`, but the `group:fga#member` relationship is missing. This is because we naively overlooked the fact that `group:fga#member` defines itself. That is, every subject who is a `member` of `group:fga` is definitely a subject of `group:fga#member`. A subject set defines itself, because it is reflexive. So our system should represent this and reflect this natively in our APIs, and what we should have had for the G2G index as a side effect of this property is the following:
+
+| object_type | object_id | relation | subject_type | subject_id | subject_relation |
+|-------------|-----------|----------|--------------|------------|------------------|
+| group       | eng       | member   | group        | fga        | member           |
+| group       | fga       | member   | group        | fga        | member           |
+
+Then our join would return:
+
+| object_type | object_id | relation | subject_type | subject_id |
+|-------------|-----------|----------|--------------|------------|
+| group       | eng       | member   | employee     | jon        |
+| group       | fga       | member   | employee     | jon        |
+
+# What it is
+[what-it-is]: #what-it-is
+
+We are introducing the notion of "implicit" tuples, i.e. tuples that were never explicitly written in the system but that will be accepted as true facts. These facts stem from applying equivalence of sets to the authorization model.
```

Comment:
I fear that if we view this change as a notion of implicit tuples, then that is misleading and raises questions like those that @rhamzeh mentioned. I think this is a misleading way to look at this change. 

We're not introducing an implicit tuple, and I apologize for even suggesting that in the first place in our initial discussions about this. The paragraph regarding "Correctness" in the Motivation section above is the perfect motivation and definition for what this work is.What we're definitively introducing is simply the notion that a userset defines itself because any subject of a set is by definition a subject of that same set. It is reflexive.

The whole point is that, without any tuples of any form, these set invariants should always hold true:
1. A userset `U` is always equal to itself `U`, because any element of the set U is most definitely an element of the same set U.
2. A userset `Ua` is a subset of some userset `U` if every element of `Ua` would also be an element of `U`. For example, 

```
type user

type document
  relations
    define editor: [user]
    define viewer: [user] or editor
```
Every element/subject of `document#editor` is also an element/subject of `document#viewer`. In this example `document#viewer` is set `U` and `document#editor` is set `Ua`

As long as the two invariants above hold true in every FGA API, then we're respecting the basic invariants of sets that we'd expect, and these invariants are independent of the type restrictions, tuples, etc..

Reply:
Ok, so instead of talking about implicit tuples, I will talk about invariants

Reply:
Thank you